### PR TITLE
Adds aria-label to the search button, as accessibility enhancement

### PR DESCRIPTION
**Pull request by Sisanu**
## Description
Adds aria-label to the search button, as accessibility enhancement.

## How has this been tested?
Tested with Lighthouse and the accessibility score went up.

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

**Comment by aristath** [comment]
I can see that before this PR, the `<button>` contained just the `<svg>` without a label.
Using `aria-label` fixes the issue, however, I'm not certain if this would be the recommended method. 🤔
I think we have 2 choices here: Either use the `aria-label` attribute, OR add in the contents of the `button` element something like `<span class="screen-reader-text">%s</span>`.

Adding the a11y tag since this will need to be reviewed 👍

**Comment by Sisanu** [comment]
Thank you @aristath for the feedback! Anything is better than no accessibility for the button.

**Comment by carolinan** [comment]
I do not know if using a screen reader text or aria-label would be best. 
<del>Since this reuses the $attributes['label'] it might also make sense to use aria-labelledby?</del> The label does not have a unique id.

**Comment by Sisanu** [comment]
> ~Since this reuses the $attributes['label'] it might also make sense to use aria-labelledby?~ The label does not have a unique id.

Maybe that's why `aria-label` is the way to go for now, without changing too much the output of the block.

**Comment by alexstine** [comment]
I think you're only other option is VisuallyHidden from wordpress/components. Not sure this would actually be better than just a simple aria-label in this case.

**Comment by alexstine** [comment]
https://accessible360.com/accessible360-blog/use-aria-label-screen-reader-text/

Looks like screen reader text is a better option and I've always opted for using screen reader text anyway. For this, it really is trivial as both should have the same effect.

@Sisanu I'll let you make the call on if you want to do the extra work with VisuallyHidden.

**Review by alexstine — APPROVED** [review]

**Comment by Sisanu** [comment]
Thank you @alexstine! I would keep it as simple as possible, and use just the `area-label` for now.

File: packages/block-library/src/search/index.php [hunk 1]
```diff
@@ -82,14 +82,15 @@ function render_block_core_search( $attributes ) {
 			$button_classes        .= ' has-icon';
 			$button_internal_markup =
 				'<svg id="search-icon" class="search-icon" viewBox="0 0 24 24" width="24" height="24">
-			        <path d="M13.5 6C10.5 6 8 8.5 8 11.5c0 1.1.3 2.1.9 3l-3.4 3 1 1.1 3.4-2.9c1 .9 2.2 1.4 3.6 1.4 3 0 5.5-2.5 5.5-5.5C19 8.5 16.5 6 13.5 6zm0 9.5c-2.2 0-4-1.8-4-4s1.8-4 4-4 4 1.8 4 4-1.8 4-4 4z"></path>
```
Comment (alexstine) [inline packages/block-library/src/search/index.php:87]: @Sisanu Just one question, why was this line changed? Just an indentation change?

Comment (Sisanu) [reply packages/block-library/src/search/index.php:87]: Yes, tabs instead of spaces, for consistency

File: packages/block-library/src/search/index.php [hunk 2]
```diff
@@ -82,14 +82,15 @@ function render_block_core_search( $attributes ) {
...
+					<path d="M13.5 6C10.5 6 8 8.5 8 11.5c0 1.1.3 2.1.9 3l-3.4 3 1 1.1 3.4-2.9c1 .9 2.2 1.4 3.6 1.4 3 0 5.5-2.5 5.5-5.5C19 8.5 16.5 6 13.5 6zm0 9.5c-2.2 0-4-1.8-4-4s1.8-4 4-4 4 1.8 4 4-1.8 4-4 4z"></path>
+				</svg>';
 		}
 
 		$button_markup = sprintf(
-			'<button type="submit" class="wp-block-search__button %s" %s>%s</button>',
+			'<button type="submit" class="wp-block-search__button %s" %s aria-label="%s">%s</button>',
 			esc_attr( $button_classes ),
 			$inline_styles['button'],
+			esc_attr( $attributes['label'] ),
```
Comment (ryelle) [inline packages/block-library/src/search/index.php]: It's possible that `label` could have HTML in it (see https://github.com/WordPress/gutenberg/pull/38649), which at least VoiceOver will read aloud - we can strip that out:

```suggestion
			esc_attr( strip_tags( $attributes['label'] ) ),
```

Comment (Sisanu) [reply packages/block-library/src/search/index.php]: Nice catch @ryelle ! I thought that `esc_attr` is already sanitizing and stripping tags. Will fix it with the next commit.

Comment (alexstine) [reply packages/block-library/src/search/index.php]: Here is the docs for this: https://developer.wordpress.org/reference/functions/esc_attr/

It does not strip tags.

Anyway, thanks @Sisanu for fixing and @ryelle for catching. I did not know tags could be sent in via this attribute. This could have ended badly.

**Comment by ryelle** [comment]
I was giving this one last test when I realized the `aria-label` was applied to every button, regardless of icon or text. The buttons that have text don't need the label, as IIRC aria-label will override the custom text someone may have set. I made a change in https://github.com/WordPress/gutenberg/pull/38136/commits/0ef7f7564a94b01f327d0d81230c31cc90c2daec to only apply the aria-label when it's an icon button (and also rebased the PR).

File: packages/block-library/src/search/index.php [hunk 3]
```diff
@@ -81,17 +82,19 @@ function render_block_core_search( $attributes ) {
 				$button_internal_markup = wp_kses_post( $attributes['buttonText'] );
 			}
 		} else {
+			$aria_label             = sprintf( 'aria-label="%s"', esc_attr( wp_strip_all_tags( $attributes['label'] ) ) );
```
Comment (alexstine) [inline packages/block-library/src/search/index.php:85]: @ryelle I am not sure using aria-label in this context should be done. Honestly, I think this whole file should be refactored to include `$attributes['button_aria_label']` that way users wouldn't be tempted to go out of their way to break accessibility. My problem with your approach is this.

```
$attributes['buttonText'] = '<span class="icon-class"></span>';
```

Now, the button is back to being inaccessible. Technically, it's not an icon button, but it still allows HTML right? Hints the need to strip tags for aria-label?

As I said, I would rather see an approach where accessibility related attributes are given and then we won't have to try to guess if users could make it inaccessible. Of course users can make it inaccessible, but the difference would be passing `$attributes['buttonText']` vs. `$attributes['button_aria_label']`.

Thoughts? Does this make sense?

**Review by alexstine — COMMENTED** [review]
A thought to add.

File: packages/block-library/src/search/index.php (hunk 3 above)
Comment (ryelle) [reply packages/block-library/src/search/index.php:85]: So, create a new attribute to pass in a button label? That does make sense, but I think it would be a good discussion to have in another issue. Overall I think getting this solution in place would improve things now, and it wouldn't prevent iterating on the label by introducing a new attribute.

Comment (alexstine) [reply packages/block-library/src/search/index.php:85]: Sure, I'm good with that. I'll make a follow-up PR once this is in.

Comment (ryelle) [reply packages/block-library/src/search/index.php:85]: Sounds good, though I'd suggest making an issue first - it'll need design input for the aria label setting UI.

**Comment by carolinan** [comment]
Please note that there is an open trac ticket that can be closed once this is added to core https://core.trac.wordpress.org/ticket/55196

**Comment by talldan** [comment]
@Sisanu Thanks for tackling this! I also made an issue about the lack of label when I noticed it recently - https://github.com/WordPress/gutenberg/issues/38647.

Looking at your PR, it looks like the problem has now been solved for the front-end output of the block.

In the editor there's some separate code for the search block's icon only button that's also missing a label:
https://github.com/WordPress/gutenberg/blob/b44f513f319f187ca96f3c5db944cf7902ca8a1a/packages/block-library/src/search/edit.js#L255-L263

Do you think a label should be added for this too? If so, is that something you'd be willing to work on?

**Comment by Sisanu** [comment]
Hi @talldan! Sure, I can take a look. Please assign me the issue.

**Comment by alexstine** [comment]
@Sisanu Looks like I can't assign you to the Issue, but here is the link.

https://github.com/WordPress/gutenberg/issues/38647

**Comment by Sisanu** [comment]
@talldan, @alexstine, it's ready for review https://github.com/WordPress/gutenberg/pull/38966

**Comment by talldan** [comment]
@Sisanu If you write something like 'Fixes #38647' or 'Closes #38647' in the PR description, there's a bot that will assign you to the issue and github will automatically close the issue when the PR is merged.
